server/src/main/java/org/elasticsearch/search/vectors/RescoreKnnVectorQuery.java

```diff
@@ -0,0 +1,160 @@
+    private final String fieldName;
+    private final byte[] byteTarget;
+    private final float[] floatTarget;
+    private final VectorSimilarityFunction vectorSimilarityFunction;
+    private final Integer k;
+    private final Query innerQuery;
+
+    private long vectorOpsCount;
+
+    public RescoreKnnVectorQuery(
+        String fieldName,
+        byte[] byteTarget,
+        VectorSimilarityFunction vectorSimilarityFunction,
+        Integer k,
+        Query innerQuery
+    ) {
+        this.fieldName = fieldName;
+        this.byteTarget = byteTarget;
+        this.floatTarget = null;
+        this.vectorSimilarityFunction = vectorSimilarityFunction;
+        this.k = k;
+        this.innerQuery = innerQuery;
+    }
+
+    public RescoreKnnVectorQuery(
+        String fieldName,
+        float[] floatTarget,
+        VectorSimilarityFunction vectorSimilarityFunction,
+        Integer k,
+        Query innerQuery
+    ) {
+        this.fieldName = fieldName;
+        this.byteTarget = null;
+        this.floatTarget = floatTarget;
+        this.vectorSimilarityFunction = vectorSimilarityFunction;
+        this.k = k;
+        this.innerQuery = innerQuery;
+    }
+
+    @Override
+    public Query rewrite(IndexSearcher searcher) throws IOException {
+        assert byteTarget == null ^ floatTarget == null : "Either byteTarget or floatTarget must be set";
+
+        final DoubleValuesSource valueSource;
+        if (byteTarget != null) {
+            valueSource = new VectorSimilarityByteValueSource(fieldName, byteTarget, vectorSimilarityFunction);
+        } else {
+            valueSource = new VectorSimilarityFloatValueSource(fieldName, floatTarget, vectorSimilarityFunction);
+        }
+        FunctionScoreQuery functionScoreQuery = new FunctionScoreQuery(innerQuery, valueSource);
+        Query query = searcher.rewrite(functionScoreQuery);
+
+        if (k == null) {
+            // No need to calculate top k - let the request size limit the results
+            return query;
+        }
+
+        // Retrieve top k documents from the rescored query
+        TopDocs topDocs = searcher.search(query, k);
+        ScoreDoc[] scoreDocs = topDocs.scoreDocs;
+        int[] docIds = new int[scoreDocs.length];
+        float[] scores = new float[scoreDocs.length];
+        for (int i = 0; i < scoreDocs.length; i++) {
+            docIds[i] = scoreDocs[i].doc;
+            scores[i] = scoreDocs[i].score;
+        }
+
+        vectorOpsCount = scoreDocs.length;
+
+        return new KnnScoreDocQuery(docIds, scores, searcher.getIndexReader());
+    }
+
+    public Query innerQuery() {
+        return innerQuery;
+    }
+
+    public Integer k() {
+        return k;
+    }
+
+    @Override
+    public void profile(QueryProfiler queryProfiler) {
+        if (innerQuery instanceof ProfilingQuery profilingQuery) {
+            profilingQuery.profile(queryProfiler);
+        }
+        queryProfiler.addVectorOpsCount(vectorOpsCount);
```

Comment:
Are we sure this works? Do we call `profile` on the un-rewritten query? Because once rewrite is called, the query is no longer this object, but instead a `FunctionScoreQuery` or a `KnnScoreDocQuery`